light-client/src/peer_list.rs

```diff
@@ -1,156 +1,206 @@
 use crate::{
     bail,
     errors::{Error, ErrorKind},
-    supervisor::Instance,
     types::PeerId,
 };
 
 use contracts::{post, pre};
-use std::collections::{HashMap, HashSet};
+use std::collections::{BTreeSet, HashMap};
 
-/// A mapping from PeerIds to Light Client instances.
-/// Keeps track of which peer is deemed the primary peer.
+/// A generic container mapping `PeerId`s to some type `T`,
+/// which keeps track of the primary peer, witnesses, full nodes,
+/// and faulty nodes. Provides lifecycle methods to swap the primary,
+/// mark witnesses as faulty, and maintains an `invariant` for
+/// correctness.
 #[derive(Debug)]
-pub struct PeerList {
-    instances: HashMap<PeerId, Instance>,
+pub struct PeerList<T> {
+    values: HashMap<PeerId, T>,
     primary: PeerId,
-    witnesses: HashSet<PeerId>,
-    full_nodes: HashSet<PeerId>,
-    faulty_nodes: HashSet<PeerId>,
+    witnesses: BTreeSet<PeerId>,
+    full_nodes: BTreeSet<PeerId>,
+    faulty_nodes: BTreeSet<PeerId>,
 }
 
-impl PeerList {
+impl<T> PeerList<T> {
```

Comment:
Is there any constraints we want to have on `T` that would help make the construction of the `PeerList` even more safe?

Reply:
I don't see any constraints which would make it safer, but perhaps I am not fully understanding/seeing what you mean by safer in this context?

Reply:
I think what I'm getting at is partially what I said in https://github.com/informalsystems/tendermint-rs/pull/397#pullrequestreview-441029225 - given this is the `PeerList` of the light client crate, it might not be all that beneficial to generalise entirely over `T` and rather constraint it to a trait (capabilities) that indicate that whatever is held as a peer value can be used as an `Instance`, or whatever other abstraction has the right surface for other components to rely/integrate on. It is in the same direction as making the `Handle` a trait, I suspect these changes to make it easier to build up the object graph in tests and safer to construct. It's mostly food for thought and can be iterated on over time.